libraries/llvm/rts.ll

```diff
@@ -440,55 +410,95 @@ copy:
+    call void @forEachObject(%Base %objectsbase, %Sp %objectssp, %Eraser @sharePositive)
 
-done:
-    %ret = phi %Stk [%stk, %entry], [%newstk, %copy]
-    ret %Stk %ret
+    %stringsbase = extractvalue %Region %region, 2, 1
+    %stringssp = extractvalue %Region %region, 2, 0
+    call void @forEachObject(%Base %stringsbase, %Sp %stringssp, %Eraser @c_buffer_refcount_increment)
+
+    %newmem.0 = call %Mem @copyMem(%Mem %mem.0)
+    %newmem.1 = call %Mem @copyMem(%Mem %mem.1)
+    %newmem.2 = call %Mem @copyMem(%Mem %mem.2)
+
+    %region.0 = insertvalue %Region undef, %Mem %newmem.0, 0
+    %region.1 = insertvalue %Region %region.0, %Mem %newmem.1, 1
+    %region.2 = insertvalue %Region %region.1, %Mem %newmem.2, 2
 
+    ret %Region %region.2
 }
 
-define void @backupRegion(%State %state, ptr %backup) {
-    %cmp = icmp eq %State %state, null
-    br i1 %cmp, label %done, label %backupState
-
-backupState:
-    %header = load %StateHeader, %State %state
-    %next = extractvalue %StateHeader %header, 0
-    %size = extractvalue %StateHeader %header, 1
-    %statedata = getelementptr %StateHeader, %State %state, i64 1
-    call void @memcpy(ptr %backup, ptr %statedata, i64 %size)
-    %backuprest = getelementptr i8, ptr %backup, i64 %size
-    tail call void @backupRegion(%State %next, ptr %backuprest)
-    ret void
+define %Stk @uniqueStack(%Stk %stk) alwaysinline {
+
+entry:
+    %stkrc = getelementptr %StkVal, %Stk %stk, i64 0, i32 0
+    %rc = load %Rc, ptr %stkrc
+    switch %Rc %rc, label %copy [%Rc 0, label %done]
 
 done:
-    ret void
+    ret %Stk %stk
+
+copy:
+    %newoldrc = sub %Rc %rc, 1
+    store %Rc %newoldrc, ptr %stkrc
+
+    %newhead = call ptr @malloc(i64 112)
+    br label %loop
+
+loop:
+    %old = phi %Stk [%stk, %copy], [%rest, %next]
+    %newstk = phi %Stk [%newhead, %copy], [%nextnew, %next]
+
+    %stkmem = getelementptr %StkVal, %Stk %old, i64 0, i32 1
+    %stkregion = getelementptr %StkVal, %Stk %old, i64 0, i32 2
+    %stkrest = getelementptr %StkVal, %Stk %old, i64 0, i32 3
+
+    %mem = load %Mem, ptr %stkmem
+    %region = load %Region, ptr %stkregion
+    %rest = load %Stk, ptr %stkrest
+
+    %newstkrc = getelementptr %StkVal, %Stk %newstk, i64 0, i32 0
+    %newstkmem = getelementptr %StkVal, %Stk %newstk, i64 0, i32 1
+    %newstkregion = getelementptr %StkVal, %Stk %newstk, i64 0, i32 2
+    %newstkrest = getelementptr %StkVal, %Stk %newstk, i64 0, i32 3
+
+    %newmem = call %Mem @copyMem(%Mem %mem)
+
+    %newsp = extractvalue %Mem %newmem, 0
+    call fastcc void @shareFrames(%Sp %newsp)
+
+    %newregion = call %Region @copyRegion(%Region %region)
+
+    store %Rc 0, ptr %newstkrc
+    store %Mem %newmem, ptr %newstkmem
+    store %Region %newregion, ptr %newstkregion
+
+    %last = icmp eq %Stk %rest, %stk
+    br i1 %last, label %closecycle, label %next
+
+next:
+    %nextnew = call ptr @malloc(i64 112)
+    store %Stk %nextnew, ptr %newstkrest
+    br label %loop
+
+closecycle:
```

Comment:
I forgot why the meta stack is cyclic...

Reply:
The meta-stack isn't cyclic but first-class stacks are.
This representation saves us a few operations on `PushStack`/`PopStack`, because we only have to swap the heads of the meta-stack and the first-class stack.